dnsmasq/patches/320-metric-order.patch

```diff
@@ -0,0 +1,192 @@
+ 
+ /* extra flags for my_syslog, we use a couple of facilities since they are known 
+    not to occupy the same bits as priorities, no matter how syslog.h is set up. */
+@@ -519,6 +520,7 @@ struct server {
+ #ifdef HAVE_LOOP
+   u32 uid;
+ #endif
++  char metric;
+   struct server *next; 
+ };
+ 
+@@ -1265,7 +1267,8 @@ void add_update_server(int flags,
+ 		       union mysockaddr *addr,
+ 		       union mysockaddr *source_addr,
+ 		       const char *interface,
+-		       const char *domain);
++		       const char *domain,
++		       char metric);
+ void check_servers(void);
+ int enumerate_interfaces(int reset);
+ void create_wildcard_listeners(void);
+diff --git a/src/forward.c b/src/forward.c
+index 942b02d..1492e44 100644
+--- a/src/forward.c
++++ b/src/forward.c
+@@ -239,6 +239,7 @@ static int forward_query(int udpfd, union mysockaddr *udpaddr,
+   struct all_addr *addrp = NULL;
+   unsigned int flags = 0;
+   struct server *start = NULL;
++  struct server *last_server = NULL;
+ #ifdef HAVE_DNSSEC
+   void *hash = hash_questions(header, plen, daemon->namebuff);
+   int do_dnssec = 0;
+@@ -372,6 +373,14 @@ static int forward_query(int udpfd, union mysockaddr *udpaddr,
+ 	    {
+ 	      if (option_bool(OPT_ORDER))
+ 		start = daemon->servers;
++	      else if (option_bool(OPT_METRIC))
++		{
++		       forward->forwardall = 1;
++		       if(!(start = daemon->last_server))
++		          start = daemon->servers;
++		       else
++		          last_server = daemon->last_server;
++		}
+ 	      else if (!(start = daemon->last_server) ||
+ 		       daemon->forwardcount++ > FORWARD_TEST ||
+ 		       difftime(now, daemon->forwardtime) > FORWARD_TIME)
+@@ -535,6 +544,9 @@ static int forward_query(int udpfd, union mysockaddr *udpaddr,
+ 	  
+ 	  if (start == firstsentto)
+ 	    break;
++
++	  if(option_bool(OPT_METRIC) && (start->flags & SERV_FROM_RESOLV) && last_server && (last_server->flags & SERV_FROM_RESOLV) && (start->metric > last_server->metric))
++	    start = daemon->servers;
+ 	}
+       
+       if (forwarded)
+@@ -839,8 +851,13 @@ void reply_query(int fd, int family, time_t now)
+ 		break;
+ 	      }
+ 	} 
+-      if (!option_bool(OPT_ALL_SERVERS))
+-	daemon->last_server = server;
++      if (!option_bool(OPT_ALL_SERVERS) && !option_bool(OPT_METRIC))
++		daemon->last_server = server;
++      if (option_bool(OPT_METRIC) && (server->flags & SERV_FROM_RESOLV))
++		if (!daemon->last_server)
++		   daemon->last_server = server;
++		else if (server->metric < daemon->last_server->metric)
++		   daemon->last_server = server;
+     }
+  
+   /* We tried resending to this server with a smaller maximum size and got an answer.
+@@ -1109,6 +1126,7 @@ void reply_query(int fd, int family, time_t now)
+ 	  send_from(forward->fd, option_bool(OPT_NOWILD) || option_bool (OPT_CLEVERBIND), daemon->packet, nn, 
+ 		    &forward->source, &forward->dest, forward->iface);
+ 	}
++	if (!option_bool(OPT_METRIC))
+       free_frec(forward); /* cancel */
```

Comment:
It look relatively safe to not free the forward as it seems to be a garbage collector in the "get_new_frec" that free any 4*TIMEOUT forwards. One better solution is to implement a delayed "free_frec" but this is what "get_new_frec" already does.

Reply:
Garbage collector look to work
Fri Mar  9 18:50:13 2018 daemon.info dnsmasq[16763]: forward->forwardall: 2
Fri Mar  9 18:50:13 2018 daemon.info dnsmasq[16763]: forward->forwardall: 2
Fri Mar  9 18:50:33 2018 daemon.info dnsmasq[16763]: free_frec: freed
Fri Mar  9 18:50:53 2018 daemon.info dnsmasq[16763]: free_frec: freed
Fri Mar  9 18:50:53 2018 daemon.info dnsmasq[16763]: free_frec: freed